File: sdks/python/src/attest/trace.py
"""Trace construction utilities for Attest."""

from __future__ import annotations

import uuid
from typing import Any

from attest._proto.types import Step, Trace, TraceMetadata


class TraceBuilder:
    """Fluent builder for constructing Trace objects."""

    def __init__(self, agent_id: str | None = None) -> None:
        self._trace_id: str = f"trc_{uuid.uuid4().hex[:12]}"
        self._agent_id: str | None = agent_id
        self._input: dict[str, Any] | None = None
        self._steps: list[Step] = []
        self._output: dict[str, Any] | None = None
        self._metadata: TraceMetadata | None = None
        self._parent_trace_id: str | None = None

    def set_trace_id(self, trace_id: str) -> TraceBuilder:
        self._trace_id = trace_id
        return self

    def set_input(self, **kwargs: Any) -> TraceBuilder:
        self._input = kwargs
        return self

    def set_input_dict(self, input_data: dict[str, Any]) -> TraceBuilder:
        self._input = input_data
        return self

    def add_llm_call(
        self,
        name: str,
        args: dict[str, Any] | None = None,
        result: dict[str, Any] | None = None,
        metadata: dict[str, Any] | None = None,
        started_at_ms: int | None = None,
        ended_at_ms: int | None = None,
        agent_id: str | None = None,
        agent_role: str | None = None,
    ) -> TraceBuilder:
        self._steps.append(
            Step(
                type="llm_call",
                name=name,
                args=args,
                result=result,
                metadata=metadata,
                started_at_ms=started_at_ms,
                ended_at_ms=ended_at_ms,
                agent_id=agent_id,
                agent_role=agent_role,
            )
        )
        return self

    def add_tool_call(
        self,
        name: str,
        args: dict[str, Any] | None = None,
        result: dict[str, Any] | None = None,
        metadata: dict[str, Any] | None = None,
        started_at_ms: int | None = None,
        ended_at_ms: int | None = None,
        agent_id: str | None = None,
        agent_role: str | None = None,
    ) -> TraceBuilder:
        from attest.simulation._context import _active_mock_registry

        registry = _active_mock_registry.get(None)
        if registry is not None and name in registry:
            mock_result = registry[name](**(args or {}))
            if isinstance(mock_result, dict):
                result = mock_result
        self._steps.append(
            Step(
                type="tool_call",
                name=name,
                args=args,
                result=result,
                metadata=metadata,
                started_at_ms=started_at_ms,
                ended_at_ms=ended_at_ms,
                agent_id=agent_id,
                agent_role=agent_role,
            )
        )
        return self

    def add_retrieval(
        self,
        name: str,
        args: dict[str, Any] | None = None,
        result: dict[str, Any] | None = None,
        metadata: dict[str, Any] | None = None,
        started_at_ms: int | None = None,
        ended_at_ms: int | None = None,
        agent_id: str | None = None,
        agent_role: str | None = None,
    ) -> TraceBuilder:
        self._steps.append(
            Step(
                type="retrieval",
                name=name,
                args=args,
                result=result,
                metadata=metadata,
                started_at_ms=started_at_ms,
                ended_at_ms=ended_at_ms,
                agent_id=agent_id,
                agent_role=agent_role,
            )
        )
        return self

    def add_step(self, step: Step) -> TraceBuilder:
        self._steps.append(step)
        return self

    def set_output(self, **kwargs: Any) -> TraceBuilder:
        self._output = kwargs
        return self

    def set_output_dict(self, output_data: dict[str, Any]) -> TraceBuilder:
        self._output = output_data
        return self

    def set_metadata(
        self,
        total_tokens: int | None = None,
        cost_usd: float | None = None,
        latency_ms: int | None = None,
        model: str | None = None,
        timestamp: str | None = None,
    ) -> TraceBuilder:
        self._metadata = TraceMetadata(
            total_tokens=total_tokens,
            cost_usd=cost_usd,
            latency_ms=latency_ms,
            model=model,
            timestamp=timestamp,
        )
        return self

    def set_parent_trace_id(self, parent_id: str) -> TraceBuilder:
        self._parent_trace_id = parent_id
        return self

    def build(self) -> Trace:
        if self._output is None:
            raise ValueError("Trace output is required. Call set_output() before build().")
        return Trace(
            trace_id=self._trace_id,
            output=self._output,
            schema_version=1,
            agent_id=self._agent_id,
            input=self._input,
            steps=list(self._steps),
            metadata=self._metadata,
            parent_trace_id=self._parent_trace_id,
        )

File: sdks/typescript/packages/core/src/trace.ts
import { randomUUID } from "node:crypto";
import type { Step, Trace, TraceMetadata } from "./proto/types.js";
import {
  STEP_LLM_CALL,
  STEP_TOOL_CALL,
  STEP_RETRIEVAL,
} from "./proto/constants.js";

export class TraceBuilder {
  private traceId: string;
  private agentId: string | undefined;
  private input: Record<string, unknown> | undefined;
  private readonly steps: Step[] = [];
  private output: Record<string, unknown> | undefined;
  private metadata: TraceMetadata | undefined;
  private parentTraceId: string | undefined;

  constructor(agentId?: string) {
    this.traceId = `trc_${randomUUID().replace(/-/g, "").slice(0, 12)}`;
    this.agentId = agentId;
  }

  setTraceId(traceId: string): this {
    this.traceId = traceId;
    return this;
  }

  setInput(input: Record<string, unknown>): this {
    this.input = input;
    return this;
  }

  addLlmCall(
    name: string,
    options?: {
      args?: Record<string, unknown>;
      result?: Record<string, unknown>;
      metadata?: Record<string, unknown>;
    },
  ): this {
    this.steps.push({
      type: STEP_LLM_CALL,
      name,
      args: options?.args,
      result: options?.result,
      metadata: options?.metadata,
    });
    return this;
  }

  addToolCall(
    name: string,
    options?: {
      args?: Record<string, unknown>;
      result?: Record<string, unknown>;
      metadata?: Record<string, unknown>;
    },
  ): this {
    this.steps.push({
      type: STEP_TOOL_CALL,
      name,
      args: options?.args,
      result: options?.result,
      metadata: options?.metadata,
    });
    return this;
  }

  addRetrieval(
    name: string,
    options?: {
      args?: Record<string, unknown>;
      result?: Record<string, unknown>;
      metadata?: Record<string, unknown>;
    },
  ): this {
    this.steps.push({
      type: STEP_RETRIEVAL,
      name,
      args: options?.args,
      result: options?.result,
      metadata: options?.metadata,
    });
    return this;
  }

  addStep(step: Step): this {
    this.steps.push(step);
    return this;
  }

  setOutput(output: Record<string, unknown>): this {
    this.output = output;
    return this;
  }

  setMetadata(metadata: {
    total_tokens?: number;
    cost_usd?: number;
    latency_ms?: number;
    model?: string;
    timestamp?: string;
  }): this {
    this.metadata = metadata;
    return this;
  }

  getTraceId(): string {
    return this.traceId;
  }

  setParentTraceId(parentId: string): this {
    this.parentTraceId = parentId;
    return this;
  }

  build(): Trace {
    if (this.output === undefined) {
      throw new Error("Trace output is required. Call setOutput() before build().");
    }
    return {
      trace_id: this.traceId,
      output: this.output,
      schema_version: 1,
      agent_id: this.agentId,
      input: this.input,
      steps: [...this.steps],
      metadata: this.metadata,
      parent_trace_id: this.parentTraceId,
    };
  }
}
